net/speedtest/speedtest_client.go

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"net"
+	"time"
+)
+
+// StartClient dials the given address and starts a speedtest.
+// It returns any errors that come up in the tests.
+// It returns an error if the given test type isn't either download or upload.
+// If there are no errors in the test, it returns a slice of results.
+func StartClient(config TestConfig, host, port string) ([]Result, error) {
+	serverAddr, err := net.ResolveTCPAddr("tcp", host+":"+port)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := net.DialTCP("tcp", nil, serverAddr)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	switch config.Type {
+	case "download":
+		conn.SetReadBuffer(LenBufJSON + MaxLenBufData)
+		configBytes, err := marshalJSON(config)
+		if err != nil {
+			return nil, err
+		}
+		conn.Write(configBytes)
+		return downloadClient(conn, config)
+	case "upload":
+		return nil, errors.New("not implemented yet")
+	default:
+		return nil, errors.New("test type invalid. Must be either download or uplaod")
+	}
+}
+
+// readJSON reads LenBufJSON number of bytes from the connection.
+// It trims the result and attempts to unmarshal the result into the given interface.
+// The given buffer must have a capacity larger than LenBufJSON.
+func readJSON(conn *net.TCPConn, buffer []byte, dest interface{}) error {
+	if cap(buffer) < LenBufJSON {
+		return errors.New("given buffer's capacity is too small")
+	}
+	buffer = buffer[:LenBufJSON]
+	_, err := io.ReadFull(conn, buffer)
+	if err != nil {
+		return err
+	}
+
+	buffer = bytes.TrimRight(buffer, "\x00")
```

Comment:
this isn't right. you shouldn't need to remove 0 bytes off things. the io.ReadFull with a fixed JSON size isn't what you want... you want to read from conn (which should probably just be an `io.Reader` too) until you get a value. You probably want a json.NewDecoder(conn) at the caller instead, and then you can just Decode values.